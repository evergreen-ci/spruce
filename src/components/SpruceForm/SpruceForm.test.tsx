import { fireEvent, render, screen, userEvent, waitFor } from "test_utils";
import { SpruceForm, SpruceFormContainer } from ".";

describe("spruce form", () => {
  it("should render as expected", () => {
    const onChange = jest.fn();
    render(
      <SpruceFormContainer title="Just a test">
        <SpruceForm
          schema={basicForm.schema}
          formData={basicForm.formData}
          onChange={onChange}
          uiSchema={basicForm.uiSchema}
        />
      </SpruceFormContainer>
    );
    expect(screen.getByLabelText("Project Cloning Method")).toBeInTheDocument();

    expect(screen.queryByText("Username Label")).not.toBeInTheDocument();
    expect(screen.queryByDataCy("add-button")).toHaveTextContent("New User");
    expect(screen.getAllByRole("heading", { level: 3 })[1]).toHaveTextContent(
      "Manage Access"
    );
  });

  it("updating the form should trigger a callback and update the form state", async () => {
    let data = {};
    const onChange = jest.fn((x) => {
      const { formData } = x;
      data = formData;
    });
    render(
      <SpruceFormContainer title="Just a test">
        <SpruceForm
          schema={basicForm.schema}
          formData={basicForm.formData}
          onChange={onChange}
          uiSchema={basicForm.uiSchema}
        />
      </SpruceFormContainer>
    );
    fireEvent.change(screen.queryByDataCy("valid-projects-input"), {
      target: { value: "new value" },
    });
    userEvent.click(screen.queryByDataCy("add-button"));
    await waitFor(() =>
      expect(screen.queryAllByDataCy("new-user-input")).toHaveLength(2)
    );
    fireEvent.change(screen.queryAllByDataCy("new-user-input")[0], {
      target: { value: "new-user" },
    });
    // eslint-disable-next-line jest/prefer-called-with
    expect(onChange).toHaveBeenCalled();
    expect(screen.queryByDataCy("valid-projects-input")).toHaveValue(
      "new value"
    );
    expect(data).toStrictEqual({
      ...basicForm.formData,
      access: null,
      validProjects: "new value",
      users: ["new-user", "initial-user"],
    });
  });

  describe("form elements", () => {
    describe("text input", () => {
      describe("invisible errors", () => {
        it("should work with validate function", () => {
          let formErrors = {};
          const onChange = jest.fn((x) => {
            const { errors } = x;
            formErrors = errors;
          });
          const validate = jest.fn((_formData, err) => err);

          const { formData, schema, uiSchema } = textInput();
          render(
            <SpruceFormContainer title="Test for Text Input">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
                validate={validate}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-input")).toHaveValue("");

          // Invisible errors should be in the form error state but not visible on the page.
          expect(formErrors).toStrictEqual([{ stack: "textInput: invisible" }]);
          expect(screen.queryByText("invisible")).toBeNull();
        });
      });

      describe("emptyValue", () => {
        it("defaults to '' when not specified", () => {
          let data = {};
          const onChange = jest.fn((x) => {
            const { formData } = x;
            data = formData;
          });
          const { formData, schema, uiSchema } = textInput();
          render(
            <SpruceFormContainer title="Test for Text Input">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-input")).toHaveValue("");
          expect(data).toStrictEqual({
            textInput: "",
          });
        });

        it("uses provided value when specified", () => {
          let data = {};
          const onChange = jest.fn((x) => {
            const { formData } = x;
            data = formData;
          });
          const { formData, schema, uiSchema } = textInput("myEmptyValue");
          render(
            <SpruceFormContainer title="Test for Text Input">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-input"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-input")).toHaveValue(
            "myEmptyValue"
          );
          expect(data).toStrictEqual({
            textInput: "myEmptyValue",
          });
        });
      });
    });

    describe("text area", () => {
      describe("invisible errors", () => {
        it("should work with validate function", () => {
          let formErrors = {};
          const onChange = jest.fn((x) => {
            const { errors } = x;
            formErrors = errors;
          });
          const validate = jest.fn((_formData, err) => err);

          const { formData, schema, uiSchema } = textArea();
          render(
            <SpruceFormContainer title="Test for Text Area">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
                validate={validate}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-area")).toHaveValue("");

          // Invisible errors should be in the form error state but not visible on the page.
          expect(formErrors).toStrictEqual([{ stack: "textArea: invisible" }]);
          expect(screen.queryByText("invisible")).toBeNull();
        });
      });

      describe("emptyValue", () => {
        it("defaults to '' when not specified", () => {
          let data = {};
          const onChange = jest.fn((x) => {
            const { formData } = x;
            data = formData;
          });
          const { formData, schema, uiSchema } = textArea();
          render(
            <SpruceFormContainer title="Test for Text Area">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-area")).toHaveValue("");
          expect(data).toStrictEqual({
            textArea: "",
          });
        });

        it("uses provided value when specified", () => {
          let data = {};
          const onChange = jest.fn((x) => {
            const { formData } = x;
            data = formData;
          });
          const { formData, schema, uiSchema } = textArea("myEmptyValue");
          render(
            <SpruceFormContainer title="Test for Text Area">
              <SpruceForm
                schema={schema}
                formData={formData}
                onChange={onChange}
                uiSchema={uiSchema}
              />
            </SpruceFormContainer>
          );
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "new value" },
          });
          fireEvent.change(screen.queryByDataCy("text-area"), {
            target: { value: "" },
          });
          expect(screen.queryByDataCy("text-area")).toHaveValue("myEmptyValue");
          expect(data).toStrictEqual({
            textArea: "myEmptyValue",
          });
        });
      });
    });

    describe("select", () => {
      it("renders with the specified default selected", () => {
        const onChange = jest.fn();
        const { formData, schema, uiSchema } = select;
        render(
          <SpruceForm
            schema={schema}
            formData={formData}
            onChange={onChange}
            uiSchema={uiSchema}
          />
        );
        expect(screen.getByText("Vanilla")).toBeInTheDocument();
        expect(screen.queryByText("Chocolate")).not.toBeInTheDocument();
        expect(screen.queryByText("Strawberry")).not.toBeInTheDocument();
      });

      it("shows three options on click", () => {
        const onChange = jest.fn();
        const { formData, schema, uiSchema } = select;
        render(
          <SpruceForm
            schema={schema}
            formData={formData}
            onChange={onChange}
            uiSchema={uiSchema}
          />
        );
        userEvent.click(screen.queryByRole("button"));
        expect(screen.queryAllByText("Vanilla")).toHaveLength(2);
        expect(screen.getByText("Chocolate")).toBeInTheDocument();
        expect(screen.getByText("Strawberry")).toBeInTheDocument();
      });

      it("closes the menu and displays the new selected option on click", async () => {
        const onChange = jest.fn();
        const { formData, schema, uiSchema } = select;
        render(
          <SpruceForm
            schema={schema}
            formData={formData}
            onChange={onChange}
            uiSchema={uiSchema}
          />
        );
        userEvent.click(screen.queryByRole("button"));
        userEvent.click(screen.queryByText("Chocolate"));
        await waitFor(() => {
          expect(screen.queryByText("Vanilla")).not.toBeInTheDocument();
        });
        expect(screen.getByText("Chocolate")).toBeInTheDocument();
        expect(screen.queryByText("Strawberry")).not.toBeInTheDocument();
      });

      it("disables options included in enumDisabled", async () => {
        const onChange = jest.fn();
        const { formData, schema, uiSchema } = select;
        render(
          <SpruceForm
            schema={schema}
            formData={formData}
            onChange={onChange}
            uiSchema={uiSchema}
          />
        );
        userEvent.click(screen.queryByRole("button"));

        // LeafyGreen doesn't label disabled options as such, so instead of checking for a property
        // attempt to click on the disabled option and verify that the dropdown doesn't close as a result.
        userEvent.click(screen.queryByText("Strawberry"));
        await waitFor(() => {
          expect(screen.queryAllByText("Vanilla")).toHaveLength(2);
        });
        expect(screen.getByText("Chocolate")).toBeInTheDocument();
        expect(screen.getByText("Strawberry")).toBeInTheDocument();
      });
    });
  });
});

const basicForm = {
  formData: {
    cloneMethod: "legacy-ssh",
    validProjects: "spruce",
    users: ["initial-user"],
  },
  schema: {
    type: "object" as "object",
    properties: {
      cloneMethod: {
        type: "string" as "string",
        title: "Project Cloning Method",
        enum: ["legacy-ssh", "oath-token"],
        enumNames: ["Legacy SSH", "Oath Token"],
      },
      validProjects: {
        type: "string" as "string",
        title: "Valid Projects",
        placeholder: "Sample input",
      },
      access: {
        type: "null" as "null",
        title: "Manage Access",
      },
      users: {
        type: "array" as "array",
        title: "Users",
        items: {
          type: "string" as "string",
          title: "Username Label",
        },
      },
    },
  },
  uiSchema: {
    cloneMethod: {
      "ui:options": {
        label: false,
      },
    },
    validProjects: {
      "ui:widget": "textarea",
      "ui:options": {
        "data-cy": "valid-projects-input",
        label: false,
      },
    },
    access: {
      "ui:rootFieldId": "access",
      "ui:sectionTitle": true,
    },
    users: {
      "ui:addButtonText": "New User",
      "ui:data-cy": "new-user-input",
      items: {
        "ui:ariaLabelledBy": "root_access",
        "ui:data-cy": "new-user-input",
      },
    },
  },
};

const textInput = (emptyValue?: string) => ({
  formData: {},
  schema: {
    type: "object" as "object",
    properties: {
      textInput: {
        type: "string" as "string",
        title: "Text Input",
        default: "",
        minLength: 1,
      },
    },
  },
  uiSchema: {
    textInput: {
      "ui:data-cy": "text-input",
      ...(emptyValue && { "ui:emptyValue": emptyValue }),
    },
  },
});

const textArea = (emptyValue?: string) => ({
  formData: {},
  schema: {
    type: "object" as "object",
    properties: {
      textArea: {
        type: "string" as "string",
        title: "Text Area",
        default: "",
        minLength: 1,
      },
    },
  },
  uiSchema: {
    textArea: {
      "ui:widget": "textarea",
      "ui:data-cy": "text-area",
      ...(emptyValue && { "ui:emptyValue": emptyValue }),
    },
  },
});

const select = {
  formData: {},
  schema: {
    type: "object" as "object",
    properties: {
      iceCream: {
        type: "string" as "string",
        title: "Ice Cream",
        default: "vanilla",
        oneOf: [
          {
            type: "string" as "string",
            title: "Vanilla",
            enum: ["vanilla"],
          },
          {
            type: "string" as "string",
            title: "Chocolate",
            enum: ["chocolate"],
          },
          {
            type: "string" as "string",
            title: "Strawberry",
            enum: ["strawberry"],
          },
        ],
      },
    },
  },
  uiSchema: {
    iceCream: {
      "ui:enumDisabled": ["strawberry"],
    },
  },
};
