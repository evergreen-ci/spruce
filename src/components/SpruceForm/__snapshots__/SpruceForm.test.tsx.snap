// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`spruceForm should render as expected 1`] = `
.emotion-0 {
  margin: 24px 0;
}

.emotion-2 {
  margin: unset;
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #21313C;
  font-size: 24px;
  line-height: 32px;
  letter-spacing: 0px;
  font-weight: medium;
}

.emotion-3 {
  margin-bottom: 48px;
  padding: 24px;
}

.emotion-5 {
  position: relative;
  border-radius: 7px;
  -webkit-transition: border 300ms ease-in-out,box-shadow 300ms ease-in-out;
  transition: border 300ms ease-in-out,box-shadow 300ms ease-in-out;
  border: 1px solid #E7EEEC;
  box-shadow: 0 4px 10px -4px rgba(6,22,33,0.3);
  background-color: white;
  color: #21313C;
}

.emotion-6 {
  margin-bottom: 20px;
}

.emotion-8 {
  max-width: 400px;
}

.emotion-10 {
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  display: block;
  margin-bottom: 2px;
  font-weight: bold;
  color: #21313C;
  font-size: 14px;
  line-height: 16px;
}

@media only screen and (max-width: 1024px) {
  .emotion-10 {
    font-size: 16px;
    line-height: 19px;
  }
}

.emotion-11 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: 0px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-radius: 4px;
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  z-index: 0;
  background-color: #F9FBFA;
  border: 1px solid #89979B;
  box-shadow: 0px 1px 2px rgba(6, 22, 33, 0.3);
  color: #3D4F58;
  font-size: 14px;
  height: 36px;
  margin-top: 2px;
  height: 36px;
  font-size: 14px;
  width: 100%;
  color: #21313C;
  border-color: #B8C4C2;
}

.emotion-11:focus {
  outline: none;
}

.emotion-11:disabled {
  pointer-events: none;
}

.emotion-11:active,
.emotion-11:focus,
.emotion-11:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-11:focus {
  color: #3D4F58;
}

.emotion-11:hover,
.emotion-11:active {
  color: #3D4F58;
  background-color: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.3),0px 0px 0px 3px #E7EEEC;
}

.emotion-11:focus {
  background: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 124, 173, 0.4),0px 0px 0px 3px #019EE2;
}

.emotion-11>span {
  padding: 0;
}

@media only screen and (max-width: 1024px) {
  .emotion-11 {
    height: 36px;
    font-size: 16px;
  }
}

.emotion-12 {
  overflow: hidden;
  border-radius: 3px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.emotion-13 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  pointer-events: none;
  position: relative;
  z-index: 0;
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding-left: 12px;
  padding-right: 12px;
}

.emotion-14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  width: 90%;
}

.emotion-15 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  max-width: 100%;
}

.emotion-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  margin-left: auto;
}

.emotion-17 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  color: #5D6C74;
  height: 16px;
  width: 16px;
  margin-left: 6px;
}

.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-21 {
  font-size: 14px;
  font-weight: bold;
  line-height: 16px;
  padding-bottom: 4px;
  color: #3D4F58;
}

.emotion-22 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  position: relative;
  z-index: 0;
}

.emotion-22:hover>[data-leafygreen-ui="interaction-ring"] {
  box-shadow: 0 0 0 3px #E7EEEC;
}

.emotion-23 {
  width: 100%;
  height: 100%;
  font-family: inherit;
  font-size: 100%;
  width: 100%;
  min-height: 64px;
  resize: none;
  border-radius: 4px;
  margin: 0;
  padding: 10px 12px 1px 12px;
  font-size: 14px;
  font-weight: normal;
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 16px;
  z-index: 1;
  border: 1px solid;
  -webkit-transition: border-color 150ms ease-in-out;
  transition: border-color 150ms ease-in-out;
  color: #21313C;
  background-color: #FFFFFF;
  border-color: #B8C4C2;
}

.emotion-23:focus {
  outline: none;
}

.emotion-23:disabled {
  cursor: not-allowed;
}

.emotion-23:focus {
  border-color: #FFFFFF;
}

.emotion-23:disabled {
  color: #89979B;
  background-color: #E7EEEC;
}

.emotion-24 {
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: absolute;
  z-index: -1;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  pointer-events: none;
  border-radius: 4px;
}

.emotion-25 {
  margin-top: 24px;
  margin-bottom: 12px;
}

.emotion-28 {
  margin-top: 16px;
  margin-bottom: 12px;
}

.emotion-30 {
  margin: unset;
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  color: #21313C;
  font-size: 18px;
  line-height: 24px;
  letter-spacing: 0px;
}

.emotion-33 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: 0px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-radius: 4px;
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  z-index: 0;
  background-color: #F9FBFA;
  border: 1px solid #89979B;
  box-shadow: 0px 1px 2px rgba(6, 22, 33, 0.3);
  color: #3D4F58;
  font-size: 14px;
  height: 28px;
}

.emotion-33:focus {
  outline: none;
}

.emotion-33:disabled {
  pointer-events: none;
}

.emotion-33:active,
.emotion-33:focus,
.emotion-33:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-33:focus {
  color: #3D4F58;
}

.emotion-33:hover,
.emotion-33:active {
  color: #3D4F58;
  background-color: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.3),0px 0px 0px 3px #E7EEEC;
}

.emotion-33:focus {
  background: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 124, 173, 0.4),0px 0px 0px 3px #019EE2;
}

.emotion-35 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 100%;
  width: 100%;
  pointer-events: none;
  position: relative;
  z-index: 0;
  font-family: Akzidenz,'Helvetica Neue',Helvetica,Arial,sans-serif;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding-left: 12px;
  padding-right: 12px;
}

.emotion-36 {
  color: #5D6C74;
  height: 16px;
  width: 16px;
  margin-right: 6px;
}

.emotion-37 {
  margin-bottom: 24px;
  min-width: -webkit-min-content;
  min-width: -moz-min-content;
  min-width: min-content;
  width: 60%;
}

.emotion-39 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-39 .field-object {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.emotion-46 p {
  margin: 0;
}

.emotion-48 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 0;
}

.emotion-49 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  position: relative;
  z-index: 0;
  width: 100%;
}

.emotion-49:hover>[data-leafygreen-ui="interaction-ring"] {
  box-shadow: 0 0 0 3px #E7EEEC;
}

.emotion-50 {
  width: 100%;
  height: 100%;
  font-family: inherit;
  font-size: 100%;
  width: 100%;
  height: 36px;
  border-radius: 4px;
  padding-left: 12px;
  font-size: 14px;
  font-weight: normal;
  font-family: Akzidenz,‘Helvetica Neue’,Helvetica,Arial,sans-serif;
  border: 1px solid;
  -webkit-transition: border-color 150ms ease-in-out;
  transition: border-color 150ms ease-in-out;
  z-index: 1;
  outline: none;
  color: #21313C;
  background-color: #FFFFFF;
  padding-right: 12px;
  border-color: #89979B;
}

.emotion-50::-webkit-input-placeholder {
  color: #89979B;
}

.emotion-50::-moz-placeholder {
  color: #89979B;
}

.emotion-50:-ms-input-placeholder {
  color: #89979B;
}

.emotion-50::placeholder {
  color: #89979B;
}

.emotion-50:disabled {
  cursor: not-allowed;
}

.emotion-50:focus {
  border: 1px solid #FFFFFF;
}

.emotion-50:disabled {
  color: #89979B;
  background-color: #E7EEEC;
}

.emotion-50:disabled:-webkit-autofill,
.emotion-50:disabled:-webkit-autofill:hover,
.emotion-50:disabled:-webkit-autofill:focus {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  border: 1px solid #89979B;
  -webkit-text-fill-color: #89979B;
  -webkit-box-shadow: 0 0 0px 1000px #E7EEEC inset;
}

.emotion-52 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  right: 12px;
  z-index: 1;
}

.emotion-53 {
  margin-bottom: 20px;
  margin-left: 16px;
  margin-top: 20px;
}

.emotion-55 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  padding: 0;
  margin: 0;
  background-color: transparent;
  border: 0px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  border-radius: 4px;
  -webkit-transition: all 150ms ease-in-out;
  transition: all 150ms ease-in-out;
  position: relative;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  z-index: 0;
  background-color: #F9FBFA;
  border: 1px solid #89979B;
  box-shadow: 0px 1px 2px rgba(6, 22, 33, 0.3);
  color: #3D4F58;
  font-size: 14px;
  height: 36px;
}

.emotion-55:focus {
  outline: none;
}

.emotion-55:disabled {
  pointer-events: none;
}

.emotion-55:active,
.emotion-55:focus,
.emotion-55:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-55:focus {
  color: #3D4F58;
}

.emotion-55:hover,
.emotion-55:active {
  color: #3D4F58;
  background-color: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.3),0px 0px 0px 3px #E7EEEC;
}

.emotion-55:focus {
  background: #FFFFFF;
  border: 1px solid #5D6C74;
  box-shadow: 0px 4px 4px rgba(0, 124, 173, 0.4),0px 0px 0px 3px #019EE2;
}

.emotion-58 {
  color: #3D4F58;
  height: 16px;
  width: 16px;
  vertical-align: text-top;
}

<div>
  <h3
    class="emotion-0 emotion-1 emotion-2"
  >
    Just a test
  </h3>
  <div
    class="emotion-3 emotion-4 emotion-5"
  >
    <form
      class="rjsf"
      novalidate=""
    >
      <div
        class="form-group field field-object"
      >
        <fieldset
          id="root"
        >
          <div
            class="form-group field field-string"
          >
            <div
              class="emotion-6 emotion-7"
            >
              <div
                class="emotion-8 emotion-9"
              >
                <div
                  class=""
                >
                  <label
                    class="emotion-10"
                    id="select-1-label"
                  >
                    Project Cloning Method
                  </label>
                  <button
                    aria-controls="select-1-menu"
                    aria-describedby="select-1-description"
                    aria-disabled="false"
                    aria-expanded="false"
                    aria-invalid="false"
                    aria-labelledby="select-1-label"
                    class="emotion-11"
                    data-testid="leafygreen-ui-select-menubutton"
                    type="button"
                    value="legacy-ssh"
                  >
                    <div
                      class="emotion-12"
                    />
                    <div
                      class="emotion-13"
                    >
                      <div
                        class="emotion-14"
                      >
                        <div
                          class="emotion-15"
                        >
                          Legacy SSH
                        </div>
                      </div>
                      <span
                        class="emotion-16"
                      >
                        <svg
                          alt=""
                          aria-hidden="true"
                          class="emotion-17"
                          height="16"
                          role="presentation"
                          viewBox="0 0 16 16"
                          width="16"
                        >
                          <path
                            d="M8.67903 10.7962C8.45271 11.0679 8.04729 11.0679 7.82097 10.7962L4.63962 6.97649C4.3213 6.59428 4.5824 6 5.06866 6L11.4313 6C11.9176 6 12.1787 6.59428 11.8604 6.97649L8.67903 10.7962Z"
                            fill="currentColor"
                          />
                        </svg>
                      </span>
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="form-group field field-string"
          >
            <div
              class="emotion-6 emotion-7"
            >
              <div
               class="emotion-20"
              >
                <label
                  class="emotion-21"
                  for="textarea-1"
                >
                  Valid Projects
                </label>
                <div
                  class="emotion-22"
                >
                  <textarea
                    class="emotion-23"
                    data-cy="valid-projects-input"
                    id="textarea-1"
                    title="Valid Projects"
                  >
                    spruce
                  </textarea>
                  <div
                    class="emotion-24"
                    data-leafygreen-ui="interaction-ring"
                  />
                </div>
              </div>
            </div>
          </div>
          <h3
            class="emotion-25 emotion-26 emotion-2"
            id="root_access"
          >
            Manage Access
          </h3>
          <div
            class="form-group field field-null"
          />
          <div
            class="form-group field field-array"
          >
            <h6
              class="emotion-28 emotion-29 emotion-30"
              id="root_users__title"
            >
              Users
            </h6>
            <div
              class="emotion-6 emotion-7"
            >
              <button
                aria-disabled="false"
                class="emotion-33"
                data-cy="add-button"
                type="button"
              >
                <div
                  class="emotion-12"
                />
                <div
                  class="emotion-35"
                >
                  <svg
                    alt=""
                    aria-hidden="true"
                    class="emotion-36"
                    fill="none"
                    height="16"
                    role="presentation"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      clip-rule="evenodd"
                      d="M7.5 2a1 1 0 00-1 1v3.5H3a1 1 0 00-1 1v1a1 1 0 001 1h3.5V13a1 1 0 001 1h1a1 1 0 001-1V9.5H13a1 1 0 001-1v-1a1 1 0 00-1-1H9.5V3a1 1 0 00-1-1h-1z"
                      fill="currentColor"
                      fill-rule="evenodd"
                    />
                  </svg>
                  New User
                </div>
              </button>
            </div>
            <div
              class="emotion-37 emotion-38"
            >
              <div
                class="emotion-39 emotion-40"
              >
                <div
                  class="form-group field field-string"
                >
                  <div
                    class="emotion-6 emotion-7"
                  >
                    <div
                      class="emotion-8 emotion-9"
                    >
                      <div
                       class="emotion-20 emotion-46 emotion-47"
                      >
                        <div
                          class="emotion-48"
                        >
                          <div
                            class="emotion-49"
                          >
                            <input
                              aria-label="Username Label"
                              aria-labelledby="root_access"
                              autocomplete="on"
                              class="emotion-50"
                              data-cy="new-user-input"
                              id="textinput-1"
                              required=""
                              type="text"
                              value="initial-user"
                            />
                            <div
                              class="emotion-24"
                              data-leafygreen-ui="interaction-ring"
                            />
                          </div>
                          <div
                            class="emotion-52"
                            data-leafygreen-ui="icon-selector"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="emotion-53 emotion-54"
                >
                  <button
                    aria-disabled="false"
                    class="emotion-55"
                    data-cy="delete-item-button"
                    type="button"
                  >
                    <div
                      class="emotion-12"
                    />
                    <div
                      class="emotion-35"
                    >
                      <svg
                        aria-label="Trash Icon"
                        class="emotion-58"
                        fill="none"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          clip-rule="evenodd"
                          d="M5 2a1 1 0 011-1h4a1 1 0 011 1h2a1 1 0 011 1v1H2V3a1 1 0 011-1h2zm9 3H2l1.678 8.392A2 2 0 005.64 15h4.72a2 2 0 001.962-1.608L14 5z"
                          fill="currentColor"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </fieldset>
      </div>
    </form>
  </div>
</div>
`;
